### DISCUSSION TITLE
Setup virtual nodes (containers) to simulate 2 mobiles and one base station.

### DISCUSSION THREAD

**Discussion by Deu82** [InterDigitalInc/AdvantEDGE#218] [Q&A] [answered]
Hello @roymx ,

First of all, thanks for helping me out to fix the meepctl deploy issues last week. Everything is running fine and I can now load the Demo 1 example and play around with it. I would like to ask for your help once more to help me setup a test use case, where: 

I would like to setup 3 virtual nodes (containers) to simulate 2 mobiles and one base station. 

After the setup is done, I would then perform the simulations on this setup. 

Since I'm a beginner with AdvanrEDGE, I don't know how to, or where can I go on the AdvantEDGE GUI to setup my test setup.

It would really be helpful if you could point me to the right direction.

Thanks in advance!

**Comment by roymx** [comment, marked as answer]
Hello @Deu82 

Regarding the steps to get started, 
- this post may be helpful in providing higher level overview of getting started - https://github.com/InterDigitalInc/AdvantEDGE/discussions/216#discussioncomment-429279
- this documentation from the Wiki will help in scenario creation - https://github.com/InterDigitalInc/AdvantEDGE/wiki/first-scenario
- finally when creating the scenario, use this button 
![image](https://user-images.githubusercontent.com/48520190/110334436-c5107200-7ff0-11eb-9860-f3cc237bb48b.png) for specific help filling the fields

Regarding the pre-requisites mentioned in the 1st link, it would help if you could provide details on the vertical you are trying to run (what is the application, is it running in containers, etc.)

Thanks,
Mike

**Reply by Deu82** [reply]
Thanks @roymx , I will follow the steps you suggested! 
I will come back to you after I explore the documentations and would inform you more details on my application case.

Have a nice day!

**Reply by Deu82** [reply]
Hello @roymx ,

I've setup a simple scenario with two UEs (in my case, two users) and one BS (base station, POA). I'm calling the UEs as near user and far user. Please see the images below:
**Image-1: my scenario** 
![Screenshot from 2021-04-07 15-34-56](https://user-images.githubusercontent.com/44395672/113876571-25ddc680-97b8-11eb-9121-f346a92552bd.png)

**Image-2: details of my scenario**
![Screenshot from 2021-04-07 15-35-18](https://user-images.githubusercontent.com/44395672/113876624-32fab580-97b8-11eb-9885-12fc68fe405e.png)

**Image-3: details of UE, UE app and edge server configurations**
![Screenshot from 2021-04-07 15-35-38](https://user-images.githubusercontent.com/44395672/113876858-6f2e1600-97b8-11eb-9c65-bd030591810f.png)
![Screenshot from 2021-04-07 15-35-55](https://user-images.githubusercontent.com/44395672/113876877-748b6080-97b8-11eb-9c87-f8beb67eff9f.png)
![Screenshot from 2021-04-07 15-36-13](https://user-images.githubusercontent.com/44395672/113876913-7bb26e80-97b8-11eb-8eca-0542fa093e69.png)

Now, I'm having **two challenges:**

**1)** I've two users: near and far user. These are named as near-user-i and far-user-j in the scenario and their parent node is the POA named bs1 . 
Could you please inform me how to place two UEs in different distances from the POA? When I add the two UEs in the GUI, they are placed automatically at same distances from the POA (here, I call them bs1[base station]).

**2)** I'd then like to execute these actions: UE1 sends x1 bits to the bs1 with an error rate e1. 
Similarly, UE2 sends x2 bits to the bs1 with error rate e2. 
I would then compare the response time. 
Could you please inform me where and how in the GUI I can execute these actions? 
I tried the 'Trigger Event' pane like below image, but could not figure out how and where can I put the bits and error rates. Besides, I could also not put two UEs at different distances:

![Screenshot from 2021-04-07 15-53-05](https://user-images.githubusercontent.com/44395672/113878102-9802db00-97b9-11eb-880f-633a6003c911.png)

Any help would be much appreciated!

Thanks and cheers!

**Reply by roymx** [reply]
Hello @Deu82 ,

There are two ways to represent "distance":
1- Topological distance (e.g. number of hops & network characteristics of hops between point A to B)
2- Geographical distance

For Topological distance, the screenshots you provided show the topological view of the network and therefore will show the topological distance (e.g. you can count the hops) - the network characteristics will be influenced by the number of hops and the net char settings of each hop.

For Geographical distance, the map screenshot that I provided below is useful when looking at the geographical distance.

**Topological distance**
In your current scenario, both UEs are at the same topological distance from the edge node and will therefore have similar network characteristics.
To increase the topological distance you can add another PoA to your zone and set different network characteristics for the new PoA (number of hops will remain the same, network characteristics will differ)
Sending a Mobility event will allow to move the UE to the new PoA and you will be able to measure new network characteristics.
Additionally, you could add new zones with new PoA that would bring the UE topologically farther from the edge node.
So bottom line, to change the "topological distance" you need to move your UE to different network locations by sending mobility events.

**Geographical distance**
In AdvantEDGE, it is possible to set geo-location for things such as UEs and PoA, it is also possible to set signal radius for the PoA. Positioning things is space allows to have Network Characteristics that can vary automatically based on the UEs physical distance to the PoA.
Steps to do this would be as follows:
1- Setup the [Geospatial Info. System](https://github.com/InterDigitalInc/AdvantEDGE/wiki/platform-concepts#geospatial-modeling) - you will need to download a map.
2- Modify your scenario to geo-locate your UEs + PoA + set a signal radius for the PoA (this is done in the configuration of your scenario)
3- Once scenario is deployed, go in Events/Automation and tick mobility and network characteristics as shown here
![image](https://user-images.githubusercontent.com/48520190/113901804-3a609580-979d-11eb-8d90-c94e68a58927.png)

The platform will connect your UEs to the closest PoA in range (the mobility automation) and will change the UE network characteristics based on UE distance to the PoA (the net. char. automation)
![image](https://user-images.githubusercontent.com/48520190/113902215-a6db9480-979d-11eb-944f-c30ca75bbaf7.png)

One final note... the physical distance between a UE and a PoA typically affects the signal strength which in turns has more influence on the available throughput rather than latency.

Let me know if you need further assistance.

**Reply by Deu82** [reply]
Hello @roymx ,

Thanks a lot for your detailed answer. I really appreciate it. I am currently trying the first approach,i.e., topological distance. However, while trying it, I found that on the '**Execute**' tab, the _Event_ and other options are disabled. Due to this I could not send a Mobility event.Although the round button is green (meaning all systems are running!). 

![Screenshot from 2021-04-14 09-39-57](https://user-images.githubusercontent.com/44395672/114673381-6a141e00-9d06-11eb-970b-9ad60dec6896.png)


A few weeks ago, when I was working with the _demo1_ app, these options were enabled. Could you please inform me what could I do to enable them.

Once these options are enables, I would then proceed with the first suggestions you made.

**A note:** Every time I start working with AdvantEDGE, I execute these commands:
meepctl deploy dep
meepctl dockerize all
meepctl deploy core

and then 
cd ~/AdvantEDGE/examples/demo1/bin/iperf-proxy
./iperf-proxy

And, after the work is done, I delete the _dep_ and the _core_

 

Thanks again for your help!

**Reply by roymx** [reply]
Hello @Deu82 ,

No problem, glad you are experimenting with the platform.

Regarding the Note, I'll refer you to the [wiki cheat sheet](https://github.com/InterDigitalInc/AdvantEDGE/wiki/deployment-cheat-sheet#common-deployment-tasks)
- when you update the AdvantEDGE repo, you need to re-build/dockerize - do 4-5-6-7
- when you start to work without updating/changing AdvantEDGE (e.g. you already built it), you just need to do 4-7

Regarding the green LED - it means that the AdvantEDGE containers are all up and running
Regarding the event buttons not enabling, it means that something in your scenario did not deploy correctly.
- If you are not using it, I recommend using this CLI tool [K9s](https://github.com/derailed/k9s) 
- Just download one of the [pre-built executable](https://github.com/derailed/k9s/releases/tag/v0.24.7) and execute on the CLI.
- it will help you identify which pods are running/failing and allows to quickly look at logs and open a shell.

My guess is that one of the POD in your scenario is not starting. I recommend doing `kubectl describe pod <name-of-failing-pod>` to identify what is wrong.

**Reply by Deu82** [reply]
Hello @roymx ,

The event tabs are now working. I created a new scenario by following the instruction [here](https://github.com/InterDigitalInc/AdvantEDGE/wiki/first-scenario) ). My previous scenario was not working properly for reasons I could not figure out. All the pods were running, still the _event_ options was not enabled. Please see my further questions below:
 


> Hello @Deu82 ,
> 
> There are two ways to represent "distance":
> 1- Topological distance (e.g. number of hops & network characteristics of hops between point A to B)
> 2- Geographical distance
> 
> For Topological distance, the screenshots you provided show the topological view of the network and therefore will show the topological distance (e.g. you can count the hops) - the network characteristics will be influenced by the number of hops and the net char settings of each hop.
> 

Since I'd like to go for the first approach (Topological distance), this is my current setup. Here I've added another PoA (named bs2) on Zone1 and added another Zone with another PoA and edge server. I've a few followup questions, please check below the image.


![Screenshot from 2021-04-15 11-19-19](https://user-images.githubusercontent.com/44395672/114846555-1c1e1980-9ddd-11eb-8962-1e479f0a65c8.png)

I've three followup questions:

1) If I send a Mobility event to allow **faruser** (in the above image) to move to _bs2_ or _bs3_, would that mean that the hops are increased now?

2) you previously mentioned that I can set different network characteristics for the PoAs "To increase the topological distance you can add another PoA to your zone and set different network characteristics for the new PoA (number of hops will remain the same, network characteristics will differ)"

-Could you please inform me where do I change the network characteristics? And, how do I change the number of hops.

3) I'd  like to execute these actions: 

i) **nearuser** sends x1 bits to the bs1 with an **error rate** e1.
ii) **faruser** sends x2 bits to the bs1 with **error rate** e2.
I would then compare the **response time**.

Could you please inform me, how do I **set the bits** (x1 and x2) and **error rate**  in the GUI? And, I suppose the **response time** is actually the **latency**, please correct me if that's wrong. 


Thanks a lot again for your help and suggestions!

**Reply by roymx** [reply]
Hello @Deu82 

Regarding 1
- Purely speaking about "_number_" of hops - number will not necessarily change (it depends on network topology)
- The hops will definitively be different.
- If connected to `bs1` PoA, hops will be `ue -> bs1 -> fog1` so that's 2 hops
- If connected to `bs2` PoA, hops will be `ue -> bs2 -> zone1 -> edge1` so that's 3 hops
- If connected to `bs3` PoA, hops will be `ue -> bs3 -> zone2 -> edge2` so that's 3 hops but not the same hops as `bs2`
- So as you can see, the "number" of hops might be the same, but the hops themselves are different and may have different network characteristics.

Regarding 2
- You can statically configure these in your scenario or sent a NET.CHAR event to change the values at run time
- Changing the number of hops requires changing the topology as you did in #1 

Regarding 3
- Response time is the latency
- error rate is also configured in the scenario - see "packet loss" parameter
- the data (e.g. "bits") are sent by the application running in the UE... so assuming you are using iperf, you would need to change the parameters of the iperf command.

**Reply by Deu82** [reply]
@roymx ,

Thanks a lot! These explanations really help to clarify some of the things :-) 👍
